features/graph/api/src/test/java/org/opennms/netmgt/graph/api/generic/GenericEdgeTest.java

```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.graph.api.generic;
+
+import static org.junit.Assert.*;
+
+import java.util.HashMap;
+import java.util.function.Supplier;
+
+import org.junit.Test;
+
+public class GenericEdgeTest {
+
+    private final static String NAMESPACE = "our_namespace";
+
+    /**
+     * We test for the condition: "The namespace of at least one VertexRef must match the namespace of the Edge".
+     */
+    @Test
+    public void namespaceShouldBeEnforced(){
+        GenericVertexRef refWithOurNamespace1 = new GenericVertexRef(NAMESPACE, "v1");
+        GenericVertexRef refWithOurNamespace2 = new GenericVertexRef(NAMESPACE, "v2");
+        GenericVertexRef refWithForeignNamespace1 = new GenericVertexRef("I don't know you, namespace", "v3");
+        GenericVertexRef refWithForeignNamespace2 = new GenericVertexRef("I don't know you either, namespace", "v4");
+        new GenericEdge(NAMESPACE, refWithOurNamespace1, refWithOurNamespace2); // should throw no Excpetion
+        new GenericEdge(NAMESPACE, refWithOurNamespace1, refWithForeignNamespace1); // should throw no Excpetion
+        new GenericEdge(NAMESPACE, refWithForeignNamespace1, refWithOurNamespace1); // should throw no Excpetion
+        assertThrowsException(IllegalArgumentException.class, () -> new GenericEdge(NAMESPACE, refWithForeignNamespace1, refWithForeignNamespace2));
+        assertThrowsException(NullPointerException.class, () -> new GenericEdge(refWithForeignNamespace1, refWithForeignNamespace2, new HashMap<>()));
+    }
+
+    private static void assertThrowsException(Class<? extends Throwable> expectedException, Supplier<?> function) {
+        try {
```

Comment:
Can you take a look at the Junit @Rules and check if we could re-write this as a Junit Rule, as this is something we need quite often.

Reply:
I am not sure if we can use @rules as a substitute since they are annotations to test methods or classes.
An alternative would be (in preferred order) to:
#  switch to Junit 5 which has this functionality already: https://junit.org/junit5/docs/current/api/org/junit/jupiter/api/Assertions.html#assertThrows(java.lang.Class,org.junit.jupiter.api.function.Executable)
# provide the functionality as a shared class in all of the tests in OpenNMS

Reply:
I am not sure if moving to Junit 5 is reasonable at the moment. There are more pressing issues at hand. If we can simply bump the version to 5.x.y and everything still works, fine, but I doubt that would be the case. 

We could provide a similar function to `assertThrows(...)` and put it somewhere in `core/test`.

Reply:
I looked a bit into moving to Junit5. It seems it is not as easy as to just change the version number but almost. Junit 5 is split into 3 jars:
core, api and vintage. Vintage allows to rung Junit4 tests to be run without changes in the junit 5 context => we would have to change all poms that reference junit to add the 3 jars.
See also: https://www.baeldung.com/junit-5-migration

I think this is out of scope for this pull request. If you are ok I would just leave the code as is and create a ticket to move to junit 5.
within that ticket would:
* move develop to junit 5
* remove all assertThrows() methods that have been added so far.

We could time box it to say 1 day and if it turns out to be more difficult as expected stop it. @mvrueden , @j-white what do you think?

Reply:
as discussed: we will add a new ticket to evaluate the pros of Junit 5 and then decide what to do.
until then we will have a test util with the assertThrows method

Reply:
I moved it to org.opennms.core.test.OnmsAssert